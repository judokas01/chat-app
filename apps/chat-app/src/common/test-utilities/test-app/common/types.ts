import { TestInterfaceModule } from '../interface'
import { TestModule } from '../no-interface'

export type CommonTestModule = TestInterfaceModule | TestModule
