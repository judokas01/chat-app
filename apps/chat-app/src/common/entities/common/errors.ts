import { InternalServerErrorException } from '@nestjs/common'

export class RelationNotLoadedError extends InternalServerErrorException {
    constructor(path: string) {
        super({
            message: `Trying to access unloaded relation! The relation at: ${path} was not loaded!`,
            payload: { path },
        })
    }
}
