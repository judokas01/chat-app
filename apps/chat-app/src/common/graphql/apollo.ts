import { ApolloDriver, ApolloDriverConfig } from '@nestjs/apollo'
import { GraphQLModule } from '@nestjs/graphql'
import { GQLContext } from '@root/common/graphql/gql.context'

export const apolloModuleAutoSchemaGen = GraphQLModule.forRoot<ApolloDriverConfig>({
    autoSchemaFile: './libs/graphql/graphql-schema.graphql',
    context: ({ res, req: reg }: GQLContext) => ({ reg, res }),
    driver: ApolloDriver,
    installSubscriptionHandlers: true,
    playground: true,
    sortSchema: true,
})

export const apolloModuleUseTypes = GraphQLModule.forRoot<ApolloDriverConfig>({
    context: ({ res, req: reg }: GQLContext) => ({ reg, res }),
    driver: ApolloDriver,
    // includeStacktraceInErrorResponses: true,
    installSubscriptionHandlers: true,
    playground: true,
    sortSchema: true,

    subscriptions: {
        'subscriptions-transport-ws': {
            onConnect: (connectionParams: ConnectionParams) => {
                return {
                    req: {
                        headers: {
                            authorization:
                                connectionParams.Authorization ?? connectionParams.authorization,
                        },
                    },
                }
            },
        },
    },
    typePaths: ['./**/*.graphql'],
})

type ConnectionParams = { Authorization?: string; authorization?: string }
