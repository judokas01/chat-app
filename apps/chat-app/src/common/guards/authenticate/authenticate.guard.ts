import { CanActivate, Inject, Injectable } from '@nestjs/common'

export interface IAuthGuard extends CanActivate {}

export const IAuthGuard = Symbol('IAuthGuard')

@Injectable()
export class AuthGuard implements IAuthGuard {
    constructor(@Inject(IAuthGuard) private authService: IAuthGuard) {}

    canActivate = this.authService.canActivate
}
