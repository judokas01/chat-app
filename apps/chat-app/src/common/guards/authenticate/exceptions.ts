import { HttpException, HttpStatus } from '@nestjs/common'

export class AuthTokenExpiredError extends HttpException {
    constructor() {
        super('Token expired.', HttpStatus.UNAUTHORIZED)
    }
}
