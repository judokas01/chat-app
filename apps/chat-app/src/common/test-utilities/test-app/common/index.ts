import { INestApplication } from '@nestjs/common'
import { TestingModule } from '@nestjs/testing'
import { IConversationRepository } from '@root/common/repositories/conversation.repository'
import { IMessageRepository } from '@root/common/repositories/message.repository'
import { IUserRepository } from '@root/common/repositories/user.repository'
import { PrismaService } from '@root/infrastructure/prisma/prisma.service'

export const cleanDb = (module: TestingModule | INestApplication) => {
    const prisma = module.get<PrismaService>(PrismaService)
    return async () => {
        await prisma.message.deleteMany({})
        await prisma.usersConversation.deleteMany({})
        await prisma.userRenewToken.deleteMany({})
        await prisma.user.deleteMany({})
        await prisma.conversation.deleteMany({})
    }
}

export const getRepositories = (module: TestingModule | INestApplication) => ({
    conversation: module.get<IConversationRepository>(IConversationRepository),
    message: module.get<IMessageRepository>(IMessageRepository),
    user: module.get<IUserRepository>(IUserRepository),
})
