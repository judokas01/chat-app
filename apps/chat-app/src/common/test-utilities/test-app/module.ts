import { Test, TestingModule } from '@nestjs/testing'
import { PrismaService } from '@root/infrastructure/prisma/prisma.service'
import { UserPrismaRepository } from '@root/auth/repository/prisma/user.repository'
import { IUserRepository } from '@root/auth/repository/user-repository.interface'
import { JWT } from '@root/auth/common/jwt.module'
import { DynamicModule, Provider, ValidationPipe, Type } from '@nestjs/common'
import { ConfigService } from '../../config/config-service.service'
import { cleanDb, getRepositories } from './common'

export const getTestModule = async (args: {
    imports?: DynamicModule[]
    providers?: Provider[]
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    controllers?: Type<any>[]
}) => {
    const { imports, providers, controllers } = args
    const module: TestingModule = await Test.createTestingModule({
        controllers: [...(controllers ?? [])],
        imports: [JWT, ...(imports ?? [])],
        providers: [
            { provide: IUserRepository, useClass: UserPrismaRepository },
            PrismaService,
            ConfigService,
            ValidationPipe,
            ...(providers ?? []),
        ],
    }).compile()

    return { cleanDb: cleanDb(module), module, repositories: getRepositories(module) }
}

export type TestModule = Awaited<ReturnType<typeof getTestModule>>
