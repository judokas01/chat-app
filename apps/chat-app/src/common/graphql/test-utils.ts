import { Query, Mutation } from '@libs/graphql/types/graphql'
import { Response } from 'supertest'
import { expect } from 'vitest'

type NoTypename<T> = Omit<T, '__typename'>

export type QueryInputs = NoTypename<Query>
export type MutationInputs = NoTypename<Mutation>

export const createOperation = <
    T extends 'Query' | 'Mutation',
    O extends keyof (T extends 'Query' ? Omit<Query, '__typename'> : Omit<Mutation, '__typename'>),
    I extends (T extends 'Query' ? QueryInputs : MutationInputs)[O] | undefined,
>(_args: {
    type: T
    operationName: O
}): NonNullable<I> => {
    throw new Error('')
}

export const responseContainsNoErrors = (body: Response['body']) => {
    expect(hasErrors(body)).toBeFalsy()
}

const hasErrors = (body: Response['body']) => {
    if (!body) {
        return {
            errors: ['Missing data in response!'],
            hasErrors: true,
        }
    }

    if (body.errors?.length ?? 0 > 0) {
        return {
            // Cast is OK due to the condition above
            errors: body.errors as unknown[],

            hasErrors: true,
        }
    }

    return false
}
