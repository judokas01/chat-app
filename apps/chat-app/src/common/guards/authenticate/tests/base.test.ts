import { afterAll, beforeAll, beforeEach, describe, expect, it, vitest } from 'vitest'
import { JwtService } from '@nestjs/jwt'
import { ConfigService } from '@root/common/config/config-service.service'
import { GraphQLModule } from '@nestjs/graphql'
import { ApolloDriver, ApolloDriverConfig } from '@nestjs/apollo'
import { INestApplication, Module } from '@nestjs/common'
import { NestFactory } from '@nestjs/core'
import supertest from 'supertest'
import { JWTPayload } from '@root/modules/auth/common/types'
import { JWT } from '@root/modules/auth/common/jwt.module'
import { faker } from '@faker-js/faker'
import { subSeconds } from 'date-fns'
import { IAuthGuard } from '../authenticate.guard'
import { AlwaysAuthenticatedAuthenticateService } from '../services/always-authenticated-authenticate.service'
import { JwtAuthGuard } from '../services/jwt-authenticate.service'
import { GuardTestingApiGql, GuardTestingApiRest } from './test-utils'

const url = `http://localhost:3000/api`
const gqlUrl = `http://localhost:3000/graphql`

describe('YourController', () => {
    describe('jwt guard tests', () => {
        let app: INestApplication
        let service: JwtService

        beforeAll(async () => {
            @Module({
                controllers: [GuardTestingApiRest],
                imports: [
                    GraphQLModule.forRoot<ApolloDriverConfig>({
                        autoSchemaFile: true,
                        driver: ApolloDriver,
                    }),
                    JWT,
                ],
                providers: [
                    GuardTestingApiGql,
                    { provide: IAuthGuard, useClass: JwtAuthGuard },
                    ConfigService,
                ],
            })
            class TestAppMod {}

            app = await NestFactory.create(TestAppMod, { logger: false })
            service = app.get<JwtService>(JwtService)
            await app.listen(3000)
        })

        afterAll(async () => {
            await app.close()
        })

        describe('when passing valid token', () => {
            let token: string
            beforeEach(async () => {
                const jwtPayload: JWTPayload = { sub: 'someId', userName: 'someUser' }
                token = `Bearer ` + (await service.signAsync(jwtPayload))
            })

            it('should return status 200 from guarded endpoint', async () => {
                const result = await supertest(url).get('/guarded').set({ authorization: token })
                expect(result.status).toBe(200)
            })

            it('should return status 200 from EP without guard', async () => {
                const result = await supertest(url).get('/free').set({ authorization: token })
                expect(result.status).toBe(200)
            })

            it('should return result from guarded gql query  - aa', async () => {
                const result = await supertest(gqlUrl)
                    .post('')
                    .set({ authorization: token })
                    .send({ query: '{ guarded }' })

                expect(result.body).toMatchObject({ data: { guarded: 'guarded' } })
            })

            it('should return result from free gql query', async () => {
                const result = await supertest(gqlUrl)
                    .post('')
                    .send({ query: '{ free }' })
                    .set({ authorization: token })
                expect(result.body).toMatchObject({ data: { free: 'free' } })
            })
        })

        describe('when passing invalid token', () => {
            const token = faker.string.uuid()
            it('should return status 401 from guarded endpoint - aa', async () => {
                const result = await supertest(url).get('/guarded').set({ authorization: token })
                expect(result.status).toBe(401)
            })

            it('should return status 200 from EP without guard', async () => {
                const result = await supertest(url).get('/free').set({ authorization: token })
                expect(result.status).toBe(200)
            })

            it('should return result from guarded gql query', async () => {
                const result = await supertest(gqlUrl)
                    .post('')
                    .send({ query: '{ guarded }' })
                    .set({ authorization: token })

                expect(result.body).toMatchSnapshot()
            })

            it('should return result from free gql query', async () => {
                const result = await supertest(gqlUrl)
                    .post('')
                    .send({ query: '{ free }' })
                    .set({ authorization: token })
                expect(result.body).toMatchObject({ data: { free: 'free' } })
            })
        })

        describe('when passing expired token', () => {
            let token: string
            beforeEach(async () => {
                vitest.useFakeTimers()
                vitest.setSystemTime(subSeconds(new Date(), 10))
                const jwtPayload: JWTPayload = { sub: 'someId', userName: 'someUser' }
                token = `Bearer ` + (await service.signAsync(jwtPayload))
                vitest.useRealTimers()
            })

            it('should return status 401 from guarded endpoint - aa', async () => {
                const result = await supertest(url).get('/guarded').set({ authorization: token })
                expect(result.status).toBe(401)
                expect(result.body).toMatchSnapshot()
            })

            it('should return result from guarded gql query', async () => {
                const result = await supertest(gqlUrl)
                    .post('')
                    .send({ query: '{ guarded }' })
                    .set({ authorization: token })

                expect(result.body).toMatchSnapshot()
            })
        })
    })

    describe('always authorized guard tests', () => {
        let app: INestApplication
        beforeAll(async () => {
            @Module({
                controllers: [GuardTestingApiRest],
                imports: [
                    GraphQLModule.forRoot<ApolloDriverConfig>({
                        autoSchemaFile: true,
                        driver: ApolloDriver,
                    }),
                ],
                providers: [
                    GuardTestingApiGql,
                    { provide: IAuthGuard, useClass: AlwaysAuthenticatedAuthenticateService },
                ],
            })
            class TestAppMod {}

            app = await NestFactory.create(TestAppMod, { logger: false })
            await app.listen(3000)
        })

        afterAll(async () => {
            await app.close()
        })

        it('should return status 200 from guarded endpoint', async () => {
            const result = await supertest(url).get('/guarded')
            expect(result.status).toBe(200)
        })

        it('should return status 200 from EP without guard', async () => {
            const result = await supertest(url).get('/free')
            expect(result.status).toBe(200)
        })

        it('should return result from guarded gql query', async () => {
            const result = await supertest(gqlUrl).post('').send({ query: '{ guarded }' })
            expect(result.body).toMatchObject({ data: { guarded: 'guarded' } })
        })

        it('should return result from free gql query', async () => {
            const result = await supertest(gqlUrl).post('').send({ query: '{ free }' })
            expect(result.body).toMatchObject({ data: { free: 'free' } })
        })
    })
})
