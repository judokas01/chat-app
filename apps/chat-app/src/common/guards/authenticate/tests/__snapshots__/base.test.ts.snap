// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`YourController > jwt guard tests > when passing invalid token > should return result from guarded gql query 1`] = `
{
  "data": null,
  "errors": [
    {
      "extensions": {
        "code": "UNAUTHENTICATED",
        "originalError": {
          "message": "Unauthorized",
          "statusCode": 401,
        },
      },
      "locations": [
        {
          "column": 3,
          "line": 1,
        },
      ],
      "message": "Unauthorized",
      "path": [
        "guarded",
      ],
    },
  ],
}
`;
